flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/convert/DataConverter.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.dynamic.convert;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import org.apache.flink.table.data.DecimalData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.table.types.logical.ArrayType;
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.MapType;
+import org.apache.flink.table.types.logical.RowType;
+
+interface DataConverter {
+  Object convert(Object object);
+
+  static DataConverter getNullable(LogicalType sourceType, LogicalType targetType) {
+    return nullable(get(sourceType, targetType));
+  }
+
+  static DataConverter get(LogicalType sourceType, LogicalType targetType) {
+    switch (targetType.getTypeRoot()) {
+      case BOOLEAN:
+      case INTEGER:
+      case FLOAT:
+      case VARCHAR:
+      case DATE:
+      case TIME_WITHOUT_TIME_ZONE:
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      case BINARY:
+      case VARBINARY:
+        return object -> object;
+      case DOUBLE:
+        return object -> {
+          if (object instanceof Float) {
+            return ((Float) object).doubleValue();
+          } else {
+            return object;
+          }
+        };
+      case BIGINT:
+        return object -> {
+          if (object instanceof Integer) {
+            return ((Integer) object).longValue();
+          } else {
+            return object;
+          }
+        };
+      case DECIMAL:
+        return object -> {
+          DecimalType toDecimalType = (DecimalType) targetType;
+          DecimalData decimalData = (DecimalData) object;
+          if (((DecimalType) sourceType).getPrecision() == toDecimalType.getPrecision()) {
+            return object;
+          } else {
+            return DecimalData.fromBigDecimal(
+                decimalData.toBigDecimal(), toDecimalType.getPrecision(), toDecimalType.getScale());
+          }
+        };
+      case TIMESTAMP_WITHOUT_TIME_ZONE:
```

Comment:
There will be date to timestamp promotion soon: https://github.com/apache/iceberg/pull/13336